test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller.go

```diff
@@ -230,25 +299,232 @@ func (r *DockerMachinePoolReconciler) reconcileNormal(ctx context.Context, clust
+	dockerMachineList := &infrav1.DockerMachineList{}
+	labels := map[string]string{
+		clusterv1.ClusterNameLabel:     cluster.Name,
+		clusterv1.MachinePoolNameLabel: machinePool.Name,
+	}
+	if err := c.List(ctx, dockerMachineList, client.InNamespace(dockerMachinePool.Namespace), client.MatchingLabels(labels)); err != nil {
+		return nil, err
+	}
+
+	return dockerMachineList, nil
+}
+
+// DeleteOrphanedDockerMachines deletes any DockerMachines owned by the DockerMachinePool that reference an invalid providerID, i.e. not in the latest copy of the node pool instances.
+func (r *DockerMachinePoolReconciler) DeleteOrphanedDockerMachines(ctx context.Context, cluster *clusterv1.Cluster, machinePool *expv1.MachinePool, dockerMachinePool *infraexpv1.DockerMachinePool, nodePoolMachineStatuses []docker.NodePoolMachineStatus) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.V(2).Info("Deleting orphaned machines", "dockerMachinePool", dockerMachinePool.Name, "namespace", dockerMachinePool.Namespace, "nodePoolMachineStatuses", nodePoolMachineStatuses)
+	dockerMachineList, err := getDockerMachines(ctx, r.Client, *cluster, *machinePool, *dockerMachinePool)
+	if err != nil {
+		return err
+	}
+
+	log.V(2).Info("DockerMachineList kind is", "kind", dockerMachineList.GetObjectKind())
+
+	nodePoolMachineNames := map[string]struct{}{}
+	for _, nodePoolMachineStatus := range nodePoolMachineStatuses {
+		nodePoolMachineNames[nodePoolMachineStatus.Name] = struct{}{}
+	}
+
+	for i := range dockerMachineList.Items {
+		dockerMachine := &dockerMachineList.Items[i]
+		if _, ok := nodePoolMachineNames[dockerMachine.Name]; !ok {
+			machine, err := util.GetOwnerMachine(ctx, r.Client, dockerMachine.ObjectMeta)
+			// Not found doesn't return an error, so we need to check for nil.
+			if err != nil {
+				return errors.Wrapf(err, "error getting owner Machine for DockerMachine %s/%s", dockerMachine.Namespace, dockerMachine.Name)
+			}
+			if machine == nil {
+				// If MachinePool is deleted, the DockerMachine and owner Machine doesn't already exist, then it will never come online.
+				if mpDeleted := isMachinePoolDeleted(ctx, r.Client, machinePool); mpDeleted {
+					log.Info("DockerMachine is orphaned and MachinePool is deleted, deleting DockerMachine", "dockerMachine", dockerMachine.Name, "namespace", dockerMachine.Namespace)
+					if err := r.Client.Delete(ctx, dockerMachine); err != nil {
+						return errors.Wrapf(err, "failed to delete orphaned DockerMachine %s/%s", dockerMachine.Namespace, dockerMachine.Name)
+					}
+				} else { // If the MachinePool still exists, then the Machine will be created, so we need to wait for that to happen.
+					return errors.Errorf("DockerMachine %s/%s has no parent Machine, will reattempt deletion once parent Machine is present", dockerMachine.Namespace, dockerMachine.Name)
```

Comment:
this is a case where we anyways need to delete the machinepool machine, so should we return the error here>

Reply:
The intention here was to wait for the Machine to come online and delete it to avoid a race condition where the MachinePool controller might be creating the Machine as the DockerMachine is about to be deleted. What do you think?